packages/main/src/Calendar.js

```diff
@@ -356,36 +362,66 @@ class Calendar extends WebComponent {
 		oNewDate.setYear(iNewYear);
 		oNewDate.setMonth(iNewMonth);
 
+
+		if (oNewDate.getYear() < YearPicker._MIN_YEAR) {
+			return;
+		}
 		this.timestamp = oNewDate.valueOf() / 1000;
 	}
 
 	_showNextYear() {
+		if (this._calendarDate.getYear() === YearPicker._MAX_YEAR) {
+			return;
+		}
+
 		const oNewDate = this._calendarDate;
 		oNewDate.setYear(this._calendarDate.getYear() + 1);
 
 		this.timestamp = oNewDate.valueOf() / 1000;
 	}
 
 	_showPrevYear() {
+		if (this._calendarDate.getYear() === YearPicker._MIN_YEAR) {
+			return;
+		}
+
 		const oNewDate = this._calendarDate;
 		oNewDate.setYear(this._calendarDate.getYear() - 1);
 
 		this.timestamp = oNewDate.valueOf() / 1000;
 	}
 
 	_showNextPageYears() {
+		if (this._yearPicker.timestamp) {
+			const oCalDate = CalendarDate.fromTimestamp(this._yearPicker.timestamp * 1000, this._primaryCalendarType);
+			oCalDate.setMonth(0);
+			oCalDate.setDate(1);
+			oCalDate.setYear(oCalDate.getYear() + YearPicker._ITEMS_COUNT - YearPicker._MIDDLE_ITEM_INDEX);
+			if (oCalDate.getYear() > YearPicker._MAX_YEAR) {
+				return;
+			}
+		}
+
 		this._yearPicker = Object.assign({}, this._yearPicker, {
-			// add 20 years to the timestamp of the monthpicker
-			timestamp: this._yearPicker.timestamp + (31536000 * 20),
+			timestamp: this._yearPicker.timestamp + (31536000 * YearPicker._ITEMS_COUNT),
 		});
 
 		this._isShiftingYears = true;
 	}
 
 	_showPrevPageYears() {
+		if (this._yearPicker.timestamp) {
```

Comment:
These blocks (the one highlighted here) and above seem very similar,  perhaps you can extract and reuse code.

Reply:
done